### PR TITLE
gc stale peers sending pre-vote messages.

### LINKED ISSUE #3366: Peers sending stale pre-vote messages should be gc
## Bug Report

**What version of Rust are you using?**
rustc 1.29.0-nightly (4f3c7a472 2018-07-17)

**What did you do?**
Running a TiKV cluster, 5 nodes, 3 replicas. After a while, we can see this log on 4 nodes:
```
node 1, 2 and 3
2018/07/30 19:50:17.875 INFO store.rs:1006: [region 772] raft message MsgRequestPreVote is stale, current conf_ver: 14 version: 19, ignore it
2018/07/30 19:50:33.274 INFO store.rs:1006: [region 772] raft message MsgRequestPreVote is stale, current conf_ver: 14 version: 19, ignore it
2018/07/30 19:50:48.675 INFO store.rs:1006: [region 772] raft message MsgRequestPreVote is stale, current conf_ver: 14 version: 19, ignore it
```

```
However on node4:
2018/07/30 19:51:50.248 INFO <unknown>:865: [region 772] 775 became pre-candidate at term 7
2018/07/30 19:51:50.248 INFO <unknown>:950: [region 772] 775 received MsgRequestPreVoteResponse from 775 at term 7
2018/07/30 19:51:50.248 INFO <unknown>:927: [region 772] 775 [logterm: 7, index: 9] sent MsgRequestPreVote request to 884 at term 7
2018/07/30 19:51:50.248 INFO <unknown>:927: [region 772] 775 [logterm: 7, index: 9] sent MsgRequestPreVote request to 773 at term 7
2018/07/30 19:51:50.248 INFO <unknown>:927: [region 772] 775 [logterm: 7, index: 9] sent MsgRequestPreVote request to 774 at term 7
```

**What did you expect to see?**
The stale peer should be gc.

**What did you see instead?**
The stale peer can't be gc.

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Stale peers could send pre-vote messages to other peers. After leader receives these messages
with stale epoch, it knows the source peer can be gc. But in the currently code the check is only on vote messages, not include pre-vote messages. I think it's a bug. This PR fix this. 

## What are the type of the changes? (mandatory)
bug fix and test case.

## How has this PR been tested? (mandatory)
make dev

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

Close #3366 .

### CODE REVIEW

File: tests/raftstore_cases/test_conf_change.rs [hunk 1]
```diff
@@ -799,6 +800,39 @@ fn test_learner_with_slow_snapshot() {
...
+    pd_client.disable_default_operator();
+
+    let r1 = cluster.run_conf_change();
+    cluster.must_put(b"k1", b"v1");
+
+    pd_client.must_add_peer(r1, new_peer(2, 2));
+    pd_client.must_add_peer(r1, new_peer(3, 3));
+    must_get_equal(&cluster.get_engine(3), b"k1", b"v1");
+
+    // replace peer 3 with peer 4 while peer 3 is isoluted.
```
Comment (siddontang) [inline tests/raftstore_cases/test_conf_change.rs]: isolated

File: tests/raftstore_cases/test_conf_change.rs [hunk 2]
```diff
@@ -799,6 +800,39 @@ fn test_learner_with_slow_snapshot() {
     assert!(count.load(Ordering::SeqCst) > 0);
 }
 
+fn test_stale_peer<T: Simulator>(cluster: &mut Cluster<T>) {
+    ::util::ci_setup();
```
Comment (siddontang) [inline tests/raftstore_cases/test_conf_change.rs]: why ci setup here?

**Comment by siddontang** [comment]
ping @hicqu

File: tests/raftstore_cases/test_conf_change.rs (hunk 2 above)
Comment (hicqu) [reply tests/raftstore_cases/test_conf_change.rs]: sorry, it's for debug.

**Comment by siddontang** [comment]
PTAL @Hoverbear

**Comment by hicqu** [comment]
ping @Hoverbear @BusyJay

**Comment by BusyJay** [comment]
We also need to consider Prevote when replicating peer from remote.

**Comment by hicqu** [comment]
@BusyJay got it. The PR is merged, so I will send another one to fix this.

**Pull request merged**